### DISCUSSION TITLE
Message serialization problems

### DISCUSSION THREAD

**Discussion by SirSilver** [ag2ai/faststream#857] [Q&A] [answered]
JSON serialization example provided in docs works as expected:
```python
@pytest.mark.asyncio()
async def test_json():
    broker = fskafka.KafkaBroker()

    @broker.subscriber("test-topic")
    async def handle(name: str, user_id: int):
        assert name == "John"
        assert user_id == 1

    async with fskafka.TestKafkaBroker(broker) as br:
        await br.publish({"name": "John", "user_id": 1}, topic="test-topic")
```

But publishing dict with single key raises `pydantic.ValidationError`
```python
async def test_json():
    broker = fskafka.KafkaBroker()

    @broker.subscriber("test-topic")
    async def handle(name: str):
        assert name == "John"

    async with fskafka.TestKafkaBroker(broker) as br:
        await br.publish({"name": "John"}, topic="test-topic")
```
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for handle
name
  Input should be a valid string [type=string_type, input_value={'name': 'John'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.4/v/string_type
```

**Comment by Lancetnik** [comment, marked as answer]
@SirSilver @kumaranvpl it's not a bug, it just an incorrect usage

When you are using the single argument as a function annotation, your message body will be serialized as this argument.

As an example:

```python
async def handler(body: int): ...
# waits any int-serializable message like b"1"
```

In this case your argument name has no matter cuz it is a total message body.

Multiple arguments is a shortcut for JSONs. In this case your message will be unpacked and serailized by various fields. Please, reade [this part](https://faststream.airt.ai/0.2/getting-started/subscription/annotation/#json-basic-serialization) carefully.

So

```python
async def handler(name: str, id: int): ...
# waits for { "name": "John", "id": 1 }
```

But in your case you are trying to consume string message, but send a Mapping.

```python
async def handler(body: str): ...
# waits for any "str", not { "body": "str" }
```

To consume single JSON, you should create a single-field pydantic model and use it for annotation.

```python
class User(BaseModel):
    name: str

async def handler(body: User): ...
# waits for { "name": "John" }
```

Btw, we can't serialize `"John"` and `{ "name": "John" }` the same way, so this behavior can't be changed

**Reply by SirSilver** [reply]
Oh, this makes sense. Thank you for the explanation.
